crates/miden-objects/src/batch/proposed_batch.rs

```diff
@@ -0,0 +1,465 @@
+            match account_updates.entry(tx.account_id()) {
+                Entry::Vacant(vacant) => {
+                    let batch_account_update = BatchAccountUpdate::new(
+                        tx.account_id(),
+                        tx.account_update().init_state_hash(),
+                        tx.account_update().final_state_hash(),
+                        vec![tx.id()],
+                        tx.account_update().details().clone(),
+                    );
+                    vacant.insert(batch_account_update);
+                },
+                Entry::Occupied(occupied) => {
+                    // This returns an error if the transactions are not correctly ordered, e.g. if
+                    // B comes before A.
+                    occupied.into_mut().merge_proven_tx(tx).map_err(|source| {
+                        BatchError::AccountUpdateError { account_id: tx.account_id(), source }
+                    })?;
+                },
+            };
+
+            // The expiration block of the batch is the minimum of all transaction's expiration
+            // block.
+            batch_expiration_block_num = batch_expiration_block_num.min(tx.expiration_block_num());
+        }
+
+        if account_updates.len() > MAX_ACCOUNTS_PER_BATCH {
+            return Err(BatchError::TooManyAccountUpdates(account_updates.len()));
+        }
+
+        // Check for duplicates in input notes.
+        // --------------------------------------------------------------------------------------------
+
+        // Check for duplicate input notes both within a transaction and across transactions.
+        // This also includes authenticated notes, as the transaction kernel doesn't check for
+        // duplicates.
+        let mut input_note_map = BTreeMap::new();
+
+        for tx in transactions.iter() {
+            for note in tx.input_notes() {
+                let nullifier = note.nullifier();
+                if let Some(first_transaction_id) = input_note_map.insert(nullifier, tx.id()) {
+                    return Err(BatchError::DuplicateInputNote {
+                        note_nullifier: nullifier,
+                        first_transaction_id,
+                        second_transaction_id: tx.id(),
+                    });
+                }
+            }
+        }
+
+        // Create input and output note set of the batch.
+        // --------------------------------------------------------------------------------------------
+
+        // Remove all output notes from the batch output note set that are consumed by transactions.
+        //
+        // This still allows transaction `A` to consume an unauthenticated note `x` and output note
+        // `y` and for transaction `B` to consume an unauthenticated note `y` and output
+        // note `x` (i.e., have a circular dependency between transactions).
+        let mut output_notes = BatchOutputNoteTracker::new(transactions.iter().map(AsRef::as_ref))?;
```

Comment:
This also makes sure that there are no duplicate output notes, right? If so, maybe worth mentioning this in the comments?